### DISCUSSION TITLE
Reset/empty FileInput filename & value

### DISCUSSION THREAD

**Discussion by DappenIT** [bokeh/bokeh#12254] [General]
I would like to discuss a possibility to change the "filename" and the "value" of a FileInput widget.
Right now it is not possible to change the values of "filename" and "value" because they are readonly. One use case is describted where it is helpful to empty these two values after an action / event.

If you reassign a new FileInput object, "filename" and "value" will be empty, but the ui is stil showing the selected filename. The needed use case is not covered.

Use case 1:
A user has the possibility to select an .csv file with values in it to upload the values to the bokeh programm to see the values in a plot. The user selects a file via a FileInput widget and starts the import with a button event. After the file is imported and shown in the plot, the FileInput widget has to be emptied so the user can select a different file. To make it clear to user that he can select a new file, it is helpful to empty the FileInput widget.

If further informations are needed, I will gladly add them.
Thank you very much.

**Comment by rcriii42** [comment]
I have the same problem. In my case I want to load data from a specially formatted excel file. I want to clear the file input field after to indicate it is ready, or in case the user selects the data another way. My workaround is to destroy and recreate the file input.

The file input declaration:
```
def upload_xl_data(attr, old, new):
    """Load data from an Excel file"""
    if file_input.filename:
        excel = io.BytesIO(base64.b64decode(file_input.value))
        pipeline = load_pump_excel.load_pipeline_from_workbook(openpyxl.load_workbook(filename=excel, data_only=True))
        update_source_data()
    else:
        print('upload_xl_data: No pipeline chosen')


file_input = FileInput(accept=".xls, .xlsm, .xlsx")
file_input.on_change('filename', upload_xl_data)

# Create the top row of widgets
top_rows = column(row(pipeline_dropdown, units_toggle, stop_button),
                  row(file_input))
```

Then the update function has the code to reset the file_input:
```
def update_source_data()
    """Update all the charts and forms after changing pipelines"""
    ...
    
   # Reset the file input
    global file_input
    file_input.remove_on_change('filename', upload_xl_data)
    del top_rows.children[1].children[-1]    # This deletes the file_input
    file_input = FileInput(accept=".xls, .xlsm, .xlsx")
    file_input.on_change('filename', upload_xl_data)
    top_rows.children[1].children.append(file_input)

```

**Reply by rcriii42** [reply]
Another reason to clear the FileInput is if I want to select the same file again, the on_change above does not trigger, so I have to hit cancel to clear the field, then select the file.

**Comment by Nmirabile777** [comment]
Hi, Does anyone know if this has been solved yet? Thanks!

**Comment by bryevdv** [comment]
I was going to say this was added in https://github.com/bokeh/bokeh/pull/13873 but after looking more closely, I think that PR only did the "directory upload" portion of the original issue. If that is the case then a new issue for the clear capability will need to be opened (TBH the original issue should have been split up in the first place since it asked for two different things)

**Reply by bryevdv** [reply]
It was added in a different PR for 3.5 

* https://github.com/bokeh/bokeh/pull/13890
